contracts/examples/nft-minter/src/nft_module.rs

```diff
@@ -0,0 +1,220 @@
+        require!(
+            payment_token == price_tag.token,
+            "Invalid token used as payment"
+        );
+        require!(
+            payment_nonce == price_tag.nonce,
+            "Invalid nonce for payment token"
+        );
+        require!(
+            payment_amount == price_tag.amount,
+            "Invalid amount as payment"
+        );
+
+        self.price_tag(nft_nonce).clear();
+
+        let nft_token_id = self.nft_token_id().get();
+        let caller = self.blockchain().get_caller();
+        self.send().direct(
+            &caller,
+            &nft_token_id,
+            nft_nonce,
+            &self.types().big_uint_from(NFT_AMOUNT),
+            &[],
+        );
+
+        let owner = self.blockchain().get_owner_address();
+        self.send()
+            .direct(&owner, &payment_token, payment_nonce, &payment_amount, &[]);
+
+        Ok(())
+    }
+
+    // views
+
+    #[allow(clippy::type_complexity)]
+    #[view(getNftPrice)]
+    fn get_nft_price(
+        &self,
+        nft_nonce: u64,
+    ) -> OptionalResult<MultiResult3<TokenIdentifier, u64, BigUint>> {
+        if self.price_tag(nft_nonce).is_empty() {
+            // NFT was already sold
+            OptionalResult::None
+        } else {
+            let price_tag = self.price_tag(nft_nonce).get();
+
+            OptionalResult::Some((price_tag.token, price_tag.nonce, price_tag.amount).into())
+        }
+    }
+
+    // callbacks
+
+    #[callback]
+    fn issue_callback(&self, #[call_result] result: AsyncCallResult<TokenIdentifier>) {
+        match result {
+            AsyncCallResult::Ok(token_id) => {
+                self.nft_token_id().set(&token_id);
+            },
+            AsyncCallResult::Err(_) => {
+                let caller = self.blockchain().get_owner_address();
+                let (returned_tokens, token_id) = self.call_value().payment_token_pair();
+                if token_id.is_egld() && returned_tokens > 0 {
+                    self.send()
+                        .direct(&caller, &token_id, 0, &returned_tokens, &[]);
+                }
+            },
+        }
+    }
+
+    // private
+
+    #[allow(clippy::too_many_arguments)]
+    fn create_nft_with_attributes<T: TopEncode>(
+        &self,
+        name: BoxedBytes,
+        royalties: BigUint,
+        attributes: T,
+        uri: BoxedBytes,
+        selling_price: BigUint,
+        token_used_as_payment: TokenIdentifier,
+        token_used_as_payment_nonce: u64,
+    ) -> SCResult<u64> {
+        self.require_token_issued()?;
+        self.require_local_roles_set()?;
+        require!(royalties <= ROYALTIES_MAX, "Royalties cannot exceed 100%");
+
+        let nft_token_id = self.nft_token_id().get();
+
+        let mut serialized_attributes = Vec::new();
+        attributes.top_encode(&mut serialized_attributes)?;
+
+        let attributes_hash = self.crypto().sha256(&serialized_attributes);
+
+        let nft_nonce = self.send().esdt_nft_create(
```

Comment:
so this is for PURE NFTs ? always value == 1 ? Can we extend for sft's as well?

Reply:
It could, but I wanted this to be as simple and as specific as possible. I'd have to add a lot of conditions otherwise, which would only make it more confusing.

Reply:
For instance, I'd have to have separate cases for issue/set roles functions.